### PR TITLE
Refactor remote settings

### LINKED ISSUE #5460: Value helps should be optional
In my opinion value helps should help the user to enter a correct value if he does not know it. If the user does know the value, he may enter it manually, which may be faster or prevent unnecessary lookups. Some examples:

1. Checkout commit
  I may want to checkout a specific commit I found somehow (for example when using `git bisect`). I cannot copy paste the commit hash into the input field. But I also cannot use the value help because there are "too many commits". I assume the checkout functionality would work just fine but I cannot use it.
  image

2. Checkout tag
  Same issue as with checkout commit.
  image

3. Normally the value help will work fine but also may have lots of values (branch list for example). If you know the exact value it feels unnecessary to have to (manually) search for it in the popup dialog. Often `POPUP_TO_DECIDE_LIST` is used which doesn't even offer search functionality.

Preventing manual entry and enforcing value helps does help with validation. But in the examples I have in mind the error handling needs to be on a higher level anyways (since branches / tags / commits may just have disappeared on the remote after refresh).

Suggested solution: Simply enable input for these fields:

image

Also the error message for value helps that failed to load could be adjusted when called from the toolbar to redirect the user to the settings page where he may enter the value manually.

### PR DESCRIPTION
**Pull request by flaiker**
Fixes #5460

This change allows the user to input HEAD manually without using the value helps that may fail to render (because of too many commits). I changed the layout using a radiobutton group as otherwise without the forced value helps it is not clear which of the option is active.

Before:
image

After:
image

Some changed behavior:
- The online/offline switch does not immediately save anymore (which I found very unintuitive). Only the save button saves the settings.
- All form elements are now rendered to save previous values (instead of storing them in instance variables and manually syncing them). Those not needed for the current selection are hidden.
- Validation is added because the value helps are not obligatory anymore. Except for checkout commit as I couldn't find a call that doesn't rely on finding the first commit in the branch and fails if there are too many commits.

Todos:
- [ ] Update documentation https://github.com/abapGit/docs.abapgit.org/pull/27
- [x] The hidden form elements seem to cause some vertical space being used for some reason. Maybe some CSS is needed to hide them _even more_.
  - Fixes in e69a7442295feced7d9b112f4f99d3f0cc5293b1
- [x] Previously there was a direct call to the persistence layer instead of using the values provided in the given repo instance. I am not sure if the given repo is always up to date, though using it exclusively makes more sense to me. Maybe the caller needs to make sure the instance is up to date (as we have no lock mechanism at all here?).
  - Seems like the repo instance is retrieved just before calling the page in `zcl_abapgit_gui_router->repository_services` so this should be fine.

Also not really sure about the `mo_repo->refresh` call on the back action. If the tab is switched before going back that doesn't trigger and if online / offline was switched the caller needs to retrieve a new instance anyways so I am not sure that really does anything at all.

### CODE REVIEW

**Comment by flaiker** [comment]
The height of the hidden elements seems to be `li` padding.

image

**Comment by mbtools** [comment]
On first view, this looks great. 👍  I will give it a test tomorrow

**Comment by g-back** [comment]
I agree, looks a lot better and is definitely more user friendly.
One issue: if you enter garbage into the commit hash (I entered 'asd'), it can dump in class `ZCL_ABAPGIT_CHUNK_LIB` in method  `render_branch_name`:
image

Probably have to add a check in the frontend for the minimum length of the commit hash.

**Comment by flaiker** [comment]
> One issue: if you enter garbage into the commit hash (I entered 'asd'), it can dump in class ZCL_ABAPGIT_CHUNK_LIB in method render_branch_name
Good catch! I added min/max=40 to the form field.

File: src/ui/zcl_abapgit_gui_page_sett_remo.clas.abap [hunk 1]
```diff
@@ -818,23 +891,62 @@ CLASS zcl_abapgit_gui_page_sett_remo IMPLEMENTATION.
...
+          lv_branch_check_error_id = c_id-tag.
+        WHEN c_head_types-pull_request.
+          lv_pull_request = io_form_data->get( c_id-pull_request ).
+          SPLIT lv_pull_request AT '@' INTO lv_url lv_branch.
+          IF lv_branch IS NOT INITIAL.
+            lv_branch = zif_abapgit_definitions=>c_git_branch-heads_prefix && lv_branch.
+          ENDIF.
+          lv_branch_check_error_id = c_id-pull_request.
+        WHEN c_head_types-commit ##NEEDED.
+          " Not implemented (needs API that doesn't rely on finding the first commit in the branch)
```
[the remark this answers is not in the document]
Comment (flaiker) [reply src/ui/zcl_abapgit_gui_page_sett_remo.clas.abap]: I added basic validation in e45d8f58f012168df3086a667f60c53110649d33

**Comment by flaiker** [comment]
> One potential upgrade: it would be nice to retrieve the last used branch when switching from PR back to normal. If I remember well the original branch is saved. But now I have to always choose it manually. Not critical anyway.

Ah that got lost. Should be back now (89b14831f5440a8deef68ab24846bb8ec66e20a5).

**Comment by flaiker** [comment]
Minor bug. If you PR is selected and active and you switch to branch but also change the URL at the same time the URL change is not saved. `switch_origin( space )` seems to overwrite the previous `set_url` call.

**Comment by flaiker** [comment]
Fixed the issue regarding switching from PR and changing the URL at the same time.

One test fails but that seems unrelated to this PR:
image

**Comment by mbtools** [comment]
I opened #5499 to fix the UT

**Comment by mbtools** [comment]
UT fixed with #5500

**Comment by mbtools** [comment]
Monday merge 😁

**Comment by sbcgua** [comment]
Very good UX improvement indeed. :+1:

**Comment by mbtools** [comment]
Plus it added the feature to dynamically adjust an HTML form based on radio selection! I will use it to refactor the "Create Tag" form, next

**Pull request merged**